### PR TITLE
Apply optimization from #44355 to retain

### PR DESCRIPTION
**Pull request by Xaeroxe**
As discussed in #44355 this PR applies a similar optimization to `Vec::retain`.  For `drain_filter`, a very similar function, this improved performance by up to 20%.

### CODE REVIEW

File: src/liballoc/vec.rs [hunk 1]
```diff
@@ -813,14 +813,19 @@ impl<T> Vec<T> {
             for i in 0..len {
                 if !f(&v[i]) {
                     del += 1;
+                    unsafe {
+                        ptr::read(&v[i]);
```
Comment (Mark-Simulacrum) [inline src/liballoc/vec.rs]: Shouldn't this be a call to drop_in_place?

**Review by Mark-Simulacrum — COMMENTED** [review]
Otherwise seems fine. I'd like another set of eyes though since I'm not really familiar with this code... r? @burntsushi

File: src/liballoc/vec.rs (hunk 1 above)
Comment (Xaeroxe) [reply src/liballoc/vec.rs]: Ah, yes it should.  I didn't know that function existed, so thank you!

**Comment by pietroalbini** [comment]
@BurntSushi this PR needs your review!

**Comment by BurntSushi** [comment]
r? @alexcrichton 

cc @Gankro

**Comment by alexcrichton** [comment]
Thanks @Xaeroxe! I wonder, could this perhaps be reimplemented now as `self.drain_filter(|x| !f(x))`?

**Comment by Xaeroxe** [comment]
Probably yes, the only advantage of this approach is that the drop_in_place call here makes the job of the optimizer easier during compilation.  That being said, reducing code duplication may outweigh that benefit.

**Comment by alexcrichton** [comment]
Ah yeah for now we've mostly been optimizing for readability rather than codegen in libstd so there's no need to worry, this for sure isn't the only case! For now though yeah let's optimize for reducing code duplication

**Comment by Xaeroxe** [comment]
@alexcrichton Done!  Thanks for your help!

**Pull request merged**